src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +218,152 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
     if (!rewardsByAccountSpecifier.length) return ''
 
     const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byValidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byValidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const selectUndelegationsAmountByValidatorAddress = memoize(
+  (
+    allUndelegationsEntries: Record<string, chainAdapters.cosmos.UndelegationEntry[]>,
+    validatorAddress: string
+  ) => {
+    return get(
+      allUndelegationsEntries,
+      validatorAddress,
+      [] as chainAdapters.cosmos.UndelegationEntry[]
+    )
+      .reduce((acc: BigNumber, undelegationEntry: chainAdapters.cosmos.UndelegationEntry) => {
+        acc = acc.plus(bnOrZero(undelegationEntry.amount))
+        return acc
+      }, bnOrZero(0))
+      .toString()
+  },
+  (
+    allUndelegationsEntries: Record<string, chainAdapters.cosmos.UndelegationEntry[]>,
+    validatorAddress: string
+  ) =>
+    get(allUndelegationsEntries, validatorAddress, [] as chainAdapters.cosmos.UndelegationEntry[])
+)
+
+export const selectRewardsAmountByValidatorAddress = memoize(
+  (allRewards: Record<string, chainAdapters.cosmos.Reward[]>, validatorAddress: string) => {
+    return get(allRewards, validatorAddress, [] as chainAdapters.cosmos.Reward[])
+      .reduce((acc: BigNumber, rewardEntry: chainAdapters.cosmos.Reward) => {
+        acc = acc.plus(bnOrZero(rewardEntry.amount))
+        return acc
+      }, bnOrZero(0))
+      .toString()
+  },
+  (allRewards: Record<string, chainAdapters.cosmos.Reward[]>, validatorAddress: string) =>
+    get(allRewards, validatorAddress, [] as chainAdapters.cosmos.Reward[])
+)
+
+export const selectTotalCryptoAmount = memoize(
+  (delegationsAmount: string, undelegationsAmount: string) => {
+    return bnOrZero(delegationsAmount).plus(bnOrZero(undelegationsAmount)).toString()
```

Comment:
@NeOMakinG There is the same kind of bug as with rewards and undelegations for this memoize. I think we could even remove this memoize for this very tiny calculation but I let you do as you wish.

Reply:
We probably could remove it as it's low perf, let's wait for the maintainer's review!

Reply:
Since this function a string, i.e a primitive that we can rely on, and the inputs themselves are just string that we `BN` and do some mathematical operations on, I'm happy with not having memoization here. `selectActiveStakingOpportunityDataByAssetId` itself has deep equal comparison for its output, meaning the output we get is referentially stable. If having a few more BN calls allows us to avoid bugs, let's go for it.